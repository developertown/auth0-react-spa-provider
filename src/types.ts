import "@auth0/auth0-spa-js";
import Auth0Client from "@auth0/auth0-spa-js/dist/typings/Auth0Client";
import { createContext, Dispatch } from "react";

import { Auth0ProviderActions } from "./actions";

export interface TokenUser {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
}

export interface RouteState {
  targetUrl: string;
}

export interface Auth0ProviderProps extends Auth0ClientOptions {
  enableDebugLogging?: boolean;
  onRedirectCallback?: (state: RouteState) => void;
}

export type ProviderStateCallbacks = {
  loginWithPopup: Auth0Client["loginWithPopup"];
  handleRedirectCallback: Auth0Client["handleRedirectCallback"];
  getIdTokenClaims: Auth0Client["getIdTokenClaims"];
  loginWithRedirect: Auth0Client["loginWithRedirect"];
  getTokenSilently: Auth0Client["getTokenSilently"];
  getTokenWithPopup: Auth0Client["getTokenWithPopup"];
  logout: Auth0Client["logout"];
};

export type LoadingState = {
  loading: true;
  isAuthenticated: false;
  popupOpen: false;
  user: undefined;
  auth0Client: undefined;
};

export type UnauthenticatedState = ProviderStateCallbacks & {
  loading: false;
  isAuthenticated: false;
  popupOpen: boolean;
  user: undefined;
  auth0Client: Auth0Client;
  handlingRedirect: false;
};

export type AuthenticatedState = ProviderStateCallbacks & {
  loading: false;
  isAuthenticated: true;
  popupOpen: boolean;
  user: TokenUser;
  auth0Client: Auth0Client;
  handlingRedirect: false;
};

export type HandlingRedirectState = (UnauthenticatedState | AuthenticatedState) & {
  handlingRedirect: true;
};

export type Auth0ProviderState = LoadingState | UnauthenticatedState | AuthenticatedState | HandlingRedirectState;

export const isLoadingState = (state: Auth0ProviderState): state is LoadingState => state.loading;

export const isUnauthenticatedState = (state: Auth0ProviderState): state is LoadingState =>
  !state.loading && !state.isAuthenticated;

export const isAuthenticatedState = (state: Auth0ProviderState): state is LoadingState =>
  !state.loading && state.isAuthenticated;

export const isLoadedState = (state: Auth0ProviderState): state is AuthenticatedState | UnauthenticatedState =>
  isUnauthenticatedState(state) || isAuthenticatedState(state);

export const isHandlingRedirectState = (state: Auth0ProviderState): state is HandlingRedirectState =>
  isLoadedState(state) && "handlingRedirect" in state;

interface Auth0ProviderValues {
  state: Auth0ProviderState;
  dispatch: Dispatch<Auth0ProviderActions>;
}

export const Auth0ProviderContext = createContext<Auth0ProviderValues | undefined>(undefined);
