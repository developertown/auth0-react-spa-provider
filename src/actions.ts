import Auth0Client from "@auth0/auth0-spa-js/dist/typings/Auth0Client";
import actionCreatorFactory from "typescript-fsa";

import { TokenUser } from "./types";

const createAction = actionCreatorFactory("AUTH0_PROVIDER");

export type Auth0LoadedParams = {
  auth0Client: Auth0Client;
  user?: TokenUser;
  handleRedirectCallback: Auth0Client["handleRedirectCallback"];
  loginWithPopup: Auth0Client["loginWithPopup"];
};
export const auth0Loaded = createAction<Auth0LoadedParams>("LOADED");

export type LoginWithPopupActionResult = {
  user?: TokenUser;
};
export const loginWithPopupAction = createAction.async<void, LoginWithPopupActionResult>("LOGIN_WITH_POPUP");

export type HandleRedirectCallbackResult = {
  user?: TokenUser;
};
export const handleRedirectCallbackAction = createAction.async<void, HandleRedirectCallbackResult>(
  "HANDLE_REDIRECT_CALLBACK",
);

export type Auth0ProviderAction =
  | ReturnType<typeof auth0Loaded>
  | ReturnType<typeof loginWithPopupAction.started>
  | ReturnType<typeof loginWithPopupAction.done>
  | ReturnType<typeof loginWithPopupAction.failed>
  | ReturnType<typeof handleRedirectCallbackAction.started>
  | ReturnType<typeof handleRedirectCallbackAction.done>
  | ReturnType<typeof handleRedirectCallbackAction.failed>;
