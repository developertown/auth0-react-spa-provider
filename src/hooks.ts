import { useCallback, useContext } from "react";

import { Auth0ProviderContext, AuthenticatedState, isLoadedState, UnauthenticatedState } from "./types";

export const useAuth0 = (): AuthenticatedState | UnauthenticatedState => {
  const context = useContext(Auth0ProviderContext);

  if (!context) {
    throw new Error("useAuth0 may only be used within the context of an <Auth0Provider />");
  }

  if (!isLoadedState(context.state)) {
    throw new Error("useAuth0 may only be used when auth0 has loaded");
  }

  return context.state;
};

export const useLog = (enableDebugLogging = false): typeof console.log => {
  return useCallback<typeof console.log>(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (message?: any, ...optionalParams: any[]) => {
      if (enableDebugLogging) {
        console.log(message, ...optionalParams);
      }
    },
    [enableDebugLogging],
  );
};
