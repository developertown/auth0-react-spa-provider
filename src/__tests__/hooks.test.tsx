import { render } from "@testing-library/react";
import React, { Dispatch } from "react";
import { Auth0ProviderAction } from "../actions";
import * as hooks from "../hooks";
import { useAuth0, useLog } from "../hooks";
import { Auth0ProviderState, Auth0ProviderContext } from "../types";

const TestComponent: React.FC<{ debug?: boolean; logMessage?: string }> = ({ debug, logMessage }) => {
  const log = useLog(debug);
  log(logMessage || "Logging!");
  return <div />;
};

describe("hooks", () => {
  describe("useLog", () => {
    let origConsole: Pick<typeof window.console, "log" | "warn" | "error">;

    beforeEach(() => {
      origConsole = {
        log: window.console.log,
        warn: window.console.warn,
        error: window.console.error,
      };

      window.console.log = jest.fn();
      window.console.warn = jest.fn();
      window.console.error = jest.fn();
    });

    afterEach(() => {
      window.console.log = origConsole.log;
      window.console.warn = origConsole.warn;
      window.console.error = origConsole.error;
    });

    it("Does not log anything if no arguments are provided", () => {
      render(<TestComponent />);
      expect(window.console.log).not.toHaveBeenCalled();
    });

    it("Does not log anything if the debug flag is set to false", () => {
      render(<TestComponent debug={false} />);
      expect(window.console.log).not.toHaveBeenCalled();
    });

    it("Logs if the debug flag is set to true", () => {
      const logString = "test";
      render(<TestComponent debug={true} logMessage={logString} />);
      expect(window.console.log).toBeCalledWith(logString);
    });
  });

  describe("useAuth0", () => {
    it("throws an error if the auth0 context is not present", async () => {
      const FakeFC: React.FC = () => {
        const auth0 = useAuth0();

        return <div data-testid="username">{auth0.user?.name}</div>;
      };

      expect(() => render(<FakeFC />)).toThrowErrorMatchingSnapshot();
    });

    it("returns the context values if the auth0 context is present", async () => {
      const mockState = ({ user: { name: "foo" } } as unknown) as Auth0ProviderState;
      const mockDispatch = {} as Dispatch<Auth0ProviderAction>;

      jest.spyOn(hooks, "useAuth0").mockImplementation(() => mockState);

      const FakeFC: React.FC = () => {
        const auth0 = useAuth0();

        return <div data-testid="username">{auth0.user?.name}</div>;
      };

      const App: React.FC = () => {
        return (
          <Auth0ProviderContext.Provider value={{ state: mockState, dispatch: mockDispatch }}>
            <FakeFC />
          </Auth0ProviderContext.Provider>
        );
      };

      const { getByTestId } = await render(<App />);
      const el = getByTestId("username");

      expect(el.textContent).toEqual(mockState.user?.name);
    });
  });
});
