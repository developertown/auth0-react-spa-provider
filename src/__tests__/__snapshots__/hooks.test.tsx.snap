// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hooks useAuth0 throws an error if the auth0 context is not present 1`] = `"useAuth0 may only be used within the context of an <Auth0Provider />"`;
